Fix #342: Async support for push_metrics

adamchalmers opened the pull request:
I tried to document the problem thoroughly on the issue itself. This is a proposed solution. I tried to share as much code as I could between `push_async` and `push`. Let me know if you'd rather I take a different approach!

Fixes: #342

mxinden commented on src/push.rs:
```diff
@@ -23,6 +24,10 @@ lazy_static! {
         .timeout(REQWEST_TIMEOUT_SEC)
         .build()
         .unwrap();
+    static ref ASYNC_HTTP_CLIENT: AsyncClient = AsyncClient::builder()
```
In my eyes defining an HTTP client as a global static is problematic, for all the reasons global variables are. While that is somewhat subjective, creating both an async client and a sync client, even though a user likely only needs one, sounds even worse.

What do people think of requiring users to create a client on demand and binding `push` as a method to such client?

adamchalmers replied:
I agree with you that the globals aren't ideal, but I feel like lazy_static is a very ergonomic solution for consumers of the library. I like the idea of the user instantiating a client, but it's got a few little usability woes:

- Users now have to explicitly depend on reqwest and familiarize themselves with it works, how to create a client, etc.
- It's a breaking change (I know this is still a 0.x crate, so breaking changes are expected, but we should try to minimize them)

What if instead, we just made async support a feature that users had to opt into? That way we could instantiate the async client if and only if the feature is enabled?

lucab replied:
I fear than adding an additional either-or feature is only making the whole crate harder to test and maintain.

adamchalmers replied:
Agreed. I've removed the feature now.

mxinden commented on src/push.rs:
```diff
@@ -60,6 +65,18 @@ pub fn push_metrics<S: BuildHasher>(
 
+/// Functions just like `push_metrics`, except the metrics are pushed
+/// asynchronously.
+pub async fn push_metrics_async<S: BuildHasher>(
```
For the sake of consistency we should also offer the remaining `pub` functions in an asynchronous fashion, e.g. `push_async_collector`.

adamchalmers replied:
Agreed, done!

mxinden reviewed (commented):
Thanks for both tracing this bug all the way down as well as proposing a bug fix.

I am likely the wrong maintainer to review this pull request as (a) I am not a user of pushing metrics to a Prometheus server and (b) I am of the opinion that pushing metrics is out of scope for this library.

With that in mind I would prefer another maintainer to get involved as well.

mxinden commented:
Fixes https://github.com/tikv/rust-prometheus/issues/342

_Properly linking back to the detailed bug report._

adamchalmers commented:
Tagging @koushiro @breeswish for review :)

mxinden commented on README.md:
```diff
@@ -25,7 +25,9 @@ This crate provides several optional components which can be enabled via [Cargo
 - `process`: Enable [process metrics](https://prometheus.io/docs/instrumenting/writing_clientlibs/#process-metrics) support.
 
-- `push`: Enable [push metrics](https://prometheus.io/docs/instrumenting/pushing/) support.
+- `push`: Enable [push metrics](https://prometheus.io/docs/instrumenting/pushing/) support with blocking network calls. Incompatible with `push-async`.
```
> Incompatible with `push-async`.

Keep in mind that Rust features are additive, thus no feature should be incompatible with another one.

https://github.com/rust-lang/cargo/issues/4328

breeswish replied:
Just an idea: For simplicity, how about leaving push-async alone and remove push, considering that from async to sync is easy (just execute an executor block_on). Discussion is welcome!

adamchalmers replied:
I'd be happy to do that! It just depends on what the maintainers of this project want -- I'm not sure how they balance breaking changes vs. updating to the latest and greatest. All my Rust projects are using async, so, selfishly, I'd be fine with that.

adamchalmers replied:
Actually I just changed my PR. Now the async functions are exported in the `push` feature. I've addressed @mxinden's concern about instantiating lazy_statics unnecessarily by moving their instantiation to their call site. Now you'll only instantiate a reqwest client if you actually use it to push metrics. I think is an improvement!

adamchalmers commented:
@mxinden @breeswish Hi! Just pinging y'all for another review :)

breeswish commented on README.md:
```diff
@@ -25,7 +25,7 @@ This crate provides several optional components which can be enabled via [Cargo
 - `process`: Enable [process metrics](https://prometheus.io/docs/instrumenting/writing_clientlibs/#process-metrics) support.
 
-- `push`: Enable [push metrics](https://prometheus.io/docs/instrumenting/pushing/) support.
+- `push`: Enable [push metrics](https://prometheus.io/docs/instrumenting/pushing/) support with blocking network calls.
```
The change here may be reverted

breeswish commented on src/lib.rs:
```diff
@@ -108,7 +108,7 @@ This library supports four features:
 * `nightly`: Enable nightly only features.
 * `process`: For collecting process info.
-* `push`: Enable push support.
+* `push`: Enable push support using blocking network calls. Incompatible with `push`.
```
We provide async push now, so maybe it can be reverted

breeswish commented on src/push.rs:
```diff
@@ -159,14 +190,50 @@ fn push<S: BuildHasher>(
+    handle_push_response(response.status(), push_url)
+}
+
+async fn push_async<S: BuildHasher>(
```
This part is very similar to `fn push`. Is it possible to use something like https://docs.rs/crate/reqwest/0.10.8/source/src/blocking/wait.rs?

adamchalmers replied:
I don't understand what the linked code has to do with duplication. I tried to reduce duplication as much as possible by breaking common code into their own functions which are called inside both async_push and push. But this can only go so far, because the underlying reqwest structs use different types.

breeswish reviewed (commented):
Thanks for the contribution! I have left a few comments.

adamchalmers commented:
Sorry I forgot about this for so many months. I've addressed most of the feedback except the one comment I didn't understand.

adamchalmers commented:
Sorry I forgot about this for so many months. I've addressed most of the feedback except the one comment I didn't understand. I'd like to get this merged, because my project is stuck on Prometheus 0.7, which means I'm stuck pulling in an insecure version of Hyper (via reqwest), which means that my work's CI blocks builds. I can always fork this repo until the PR is merged, but I'd rather get this addressed :)

adamchalmers commented:
PR is no longer necessary because I realized that Tokio 1.0 doesn't clash with reqwest::blocking!

adamchalmers commented:
Actually, the latest reqwest 0.11 and the latest tokio 1 are incompatible again. I think this PR is ultimately the best way forward, that way projects which use rust-prometheus can make their own choice of async or blocking.

WaffleLapkin commented:
What's the status of this PR? It seems odd, to not provide an async interface when using `reqwest`. Calling blocking code (especially something like an HTTP client) from async is never a good idea.

ashlineldridge commented:
Any update on this PR? Would be great to this functionality land 🙏🏼

adamchalmers commented:
I agree, it would be great for one of the maintainers to review this and tell me what is blocking merge.

adamchalmers commented:
Closing as apparently no maintainers want this reviewed. Ping me if you want it reopened.
